twins/api/http/endpoint_states_test.go

```diff
@@ -0,0 +1,256 @@
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf(queryFmt, baseURL, 0, -5),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with zero limit",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf(queryFmt, baseURL, 0, 0),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with limit greater than max",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf(queryFmt, baseURL, 0, 110),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with invalid offset",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s?offset=invalid&limit=%d", baseURL, 15),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with invalid limit",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s?offset=%d&limit=invalid", baseURL, 0),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states without offset",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?limit=%d", baseURL, 15),
+			res:    data[0:15],
+		},
+		{
+			desc:   "get a list of states without limit",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?offset=%d", baseURL, 14),
+			res:    data[14:24],
+		},
+		{
+			desc:   "get a list of states with invalid number of params",
+			auth:   token,
+			status: http.StatusBadRequest,
+			url:    fmt.Sprintf("%s%s", baseURL, "?offset=4&limit=4&limit=5&offset=5"),
+			res:    nil,
+		},
+		{
+			desc:   "get a list of states with redundant query params",
+			auth:   token,
+			status: http.StatusOK,
+			url:    fmt.Sprintf("%s?offset=%d&limit=%d&value=something", baseURL, 0, 5),
+			res:    data[0:5],
+		},
+	}
+
+	for _, tc := range cases {
+		req := testRequest{
+			client: ts.Client(),
+			method: http.MethodGet,
+			url:    tc.url,
+			token:  tc.auth,
+		}
+		res, err := req.make()
+
+		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+		var resData statesPageRes
+		json.NewDecoder(res.Body).Decode(&resData)
+		assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		assert.ElementsMatch(t, tc.res, resData.States, fmt.Sprintf("%s: expected body %v got %v", tc.desc, tc.res, resData.States))
+	}
+}
+
+func createAttribute(name, subtopic string) twins.Attribute {
+	id, _ := uuid.New().ID()
+	return twins.Attribute{
+		Name:         name,
+		Channel:      id,
+		Subtopic:     subtopic,
+		PersistState: true,
+	}
+}
+
+func createSenML(n int, bn string) []senml.Record {
+	var recs []senml.Record
+	bt := time.Now().Unix()
+	for i := 0; i < n; i++ {
+		rec := senml.Record{
+			BaseName: bn,
+			BaseTime: float64(bt),
+			Time:     float64(i),
+			Value:    nil,
+		}
+		recs = append(recs, rec)
+	}
+	return recs
+}
+
+func createMessage(attr twins.Attribute, recs []senml.Record) *messaging.Message {
+	mRecs, _ := json.Marshal(recs)
```

Comment:
Yes, but it's a helper function - not a test func with `t *testing.T` and the code guarantees correct values, since the values are generated in this package for this purpose and in this program flow - i.e. they don't come from outside.